src/nbl/video/CVKLogicalDevice.h

```diff
@@ -178,52 +802,382 @@ class CVKLogicalDevice final : public ILogicalDevice
+        vk_createInfo.flags = 0; // Todo(achal): I would need to create a IDescriptorSetLayout::SCreationParams for this
+        vk_createInfo.bindingCount = bindingCount;
+        vk_createInfo.pBindings = vk_dsLayoutBindings;
+
+        VkDescriptorSetLayout vk_dsLayout;
+        if (vkCreateDescriptorSetLayout(m_vkdev, &vk_createInfo, nullptr, &vk_dsLayout) == VK_SUCCESS)
+        {
+            return core::make_smart_refctd_ptr<CVulkanDescriptorSetLayout>(
+                core::smart_refctd_ptr<CVKLogicalDevice>(this), _begin, _end, vk_dsLayout);
+        }
+        else
+        {
+            return nullptr;
+        }
     }
 
     core::smart_refctd_ptr<IGPUPipelineLayout> createGPUPipelineLayout_impl(const asset::SPushConstantRange* const _pcRangesBegin = nullptr,
-        const asset::SPushConstantRange* const _pcRangesEnd = nullptr, core::smart_refctd_ptr<IGPUDescriptorSetLayout>&& _layout0 = nullptr,
-        core::smart_refctd_ptr<IGPUDescriptorSetLayout>&& _layout1 = nullptr, core::smart_refctd_ptr<IGPUDescriptorSetLayout>&& _layout2 = nullptr,
-        core::smart_refctd_ptr<IGPUDescriptorSetLayout>&& _layout3 = nullptr) override
+        const asset::SPushConstantRange* const _pcRangesEnd = nullptr, core::smart_refctd_ptr<IGPUDescriptorSetLayout>&& layout0 = nullptr,
+        core::smart_refctd_ptr<IGPUDescriptorSetLayout>&& layout1 = nullptr, core::smart_refctd_ptr<IGPUDescriptorSetLayout>&& layout2 = nullptr,
+        core::smart_refctd_ptr<IGPUDescriptorSetLayout>&& layout3 = nullptr) override
     {
-        return nullptr;
+        constexpr uint32_t MAX_PC_RANGE_COUNT = 100u;
+        constexpr uint32_t MAX_DESCRIPTOR_SET_LAYOUT_COUNT = 4u; // temporary max, I believe
+
+        const core::smart_refctd_ptr<IGPUDescriptorSetLayout> tmp[] = { layout0, layout1, layout2,
+            layout3 };
+
+        VkDescriptorSetLayout vk_dsLayouts[MAX_DESCRIPTOR_SET_LAYOUT_COUNT];
+        uint32_t dsLayoutCount = 0u;
+        for (uint32_t i = 0u; i < MAX_DESCRIPTOR_SET_LAYOUT_COUNT; ++i)
+        {
+            if (tmp[i] && (tmp[i]->getAPIType() == EAT_VULKAN))
+                vk_dsLayouts[dsLayoutCount++] = static_cast<const CVulkanDescriptorSetLayout*>(tmp[i].get())->getInternalObject();
+        }
+
+        const auto pcRangeCount = std::distance(_pcRangesBegin, _pcRangesEnd);
+        assert(pcRangeCount <= MAX_PC_RANGE_COUNT);
+        VkPushConstantRange vk_pushConstantRanges[MAX_PC_RANGE_COUNT];
+        for (uint32_t i = 0u; i < pcRangeCount; ++i)
+        {
+            const auto pcRange = _pcRangesBegin + i;
+
+            vk_pushConstantRanges[i].stageFlags = static_cast<VkShaderStageFlags>(pcRange->stageFlags);
+            vk_pushConstantRanges[i].offset = pcRange->offset;
+            vk_pushConstantRanges[i].size = pcRange->size;
+        }
```

Comment:
I think Nabla's push constant struct is bytewise compatible (you could check with a few asserts about member offsets, size and enum values

so maybe you could `reinterpret_cast`

Reply:
@devshgraphicsprogramming 
Do we currently get the "VkPhysicalDevice" from physical device and store it somewhere  for these kind of assertions?

I know I would use the limits (or PhysicalDeviceRayTracingProperties) for validations of users input.

Although Vulkan has guaranteed some minimums but those aren't enough. For example for RayTracing the "recursion depth"

Reply:
Here:
https://github.com/Devsh-Graphics-Programming/Nabla/blob/apis_as_plugins/include/nbl/video/IPhysicalDevice.h#L21

We need to have a talk about make optional features/extensions uniform acrosss all APIs.